### ISSUE TITLE
Crash (segmentation fault) when storing data to HDFS

### ISSUE THREAD

**Issue by filipski** [uber/petastorm#502]
The simple code to ingest images (listed below) works fine with output_url set to a local file system ('file://') but crashes with just 'Segmentation fault (core dumped)' when trying to save data to HDFS on high-availability Hadoop cluster.

It looks like some finalizing code from `materialize_dataset` fails, as the segfault is just after printing out 'DONE', so 'STAGE 3' finishes well.
All .parquet files are properly created on HDFS (with correct file size, compared with the same output on the local file system), there's also `_SUCCESS` flag file, but the only one which is missing is `_common_metadata` file. Apparently the crash happens when writing this file.

I tried increasing `spark.driver.memory` up to 8g in:
```python
spark = SparkSession.builder.config('spark.driver.memory', '2g').master('local[*]').getOrCreate()
```
I also tried limiting the CPU to 'local[1]' or 'local[2]', with no effect, it still crashes the same way.

**Folder content**
Here's what's in the output folder after the crash:
```
$ hdfs dfs -ls /data/petastorm_ingestion_tests
Found 11 items
-rw-r--r--   3 hduser hadoop          0 2020-03-05 15:04 /data/petastorm_ingestion_tests/_SUCCESS
-rw-r--r--   3 hduser hadoop    3204443 2020-03-05 15:04 /data/petastorm_ingestion_tests/part-00000-e10468ee-72d2-4a8f-b000-f63ed0a6427b-c000.parquet
-rw-r--r--   3 hduser hadoop    3267115 2020-03-05 15:03 /data/petastorm_ingestion_tests/part-00001-e10468ee-72d2-4a8f-b000-f63ed0a6427b-c000.parquet
-rw-r--r--   3 hduser hadoop    4943022 2020-03-05 15:03 /data/petastorm_ingestion_tests/part-00002-e10468ee-72d2-4a8f-b000-f63ed0a6427b-c000.parquet
-rw-r--r--   3 hduser hadoop    4937730 2020-03-05 15:04 /data/petastorm_ingestion_tests/part-00003-e10468ee-72d2-4a8f-b000-f63ed0a6427b-c000.parquet
-rw-r--r--   3 hduser hadoop    4911543 2020-03-05 15:03 /data/petastorm_ingestion_tests/part-00004-e10468ee-72d2-4a8f-b000-f63ed0a6427b-c000.parquet
-rw-r--r--   3 hduser hadoop    3132042 2020-03-05 15:04 /data/petastorm_ingestion_tests/part-00005-e10468ee-72d2-4a8f-b000-f63ed0a6427b-c000.parquet
-rw-r--r--   3 hduser hadoop    4816829 2020-03-05 15:04 /data/petastorm_ingestion_tests/part-00006-e10468ee-72d2-4a8f-b000-f63ed0a6427b-c000.parquet
-rw-r--r--   3 hduser hadoop    4835624 2020-03-05 15:04 /data/petastorm_ingestion_tests/part-00007-e10468ee-72d2-4a8f-b000-f63ed0a6427b-c000.parquet
-rw-r--r--   3 hduser hadoop    3187349 2020-03-05 15:03 /data/petastorm_ingestion_tests/part-00008-e10468ee-72d2-4a8f-b000-f63ed0a6427b-c000.parquet
-rw-r--r--   3 hduser hadoop    6273385 2020-03-05 15:04 /data/petastorm_ingestion_tests/part-00009-e10468ee-72d2-4a8f-b000-f63ed0a6427b-c000.parquet
```

And when I pass `use_summary_metadata=True` to `materialize_dataset()` it still crashes but creates the following files:

```
$ hdfs dfs -ls /data/petastorm_ingestion_tests
Found 13 items
-rw-r--r--   3 hduser hadoop          0 2020-03-05 15:01 /data/petastorm_ingestion_tests/_SUCCESS
-rw-r--r--   3 hduser hadoop        351 2020-03-05 15:01 /data/petastorm_ingestion_tests/_common_metadata
-rw-r--r--   3 hduser hadoop       3950 2020-03-05 15:01 /data/petastorm_ingestion_tests/_metadata
-rw-r--r--   3 hduser hadoop    3204443 2020-03-05 15:00 /data/petastorm_ingestion_tests/part-00000-daa9b6fc-7cf1-4404-aad5-2870e6e85c76-c000.parquet
-rw-r--r--   3 hduser hadoop    3267115 2020-03-05 15:00 /data/petastorm_ingestion_tests/part-00001-daa9b6fc-7cf1-4404-aad5-2870e6e85c76-c000.parquet
-rw-r--r--   3 hduser hadoop    4943022 2020-03-05 15:01 /data/petastorm_ingestion_tests/part-00002-daa9b6fc-7cf1-4404-aad5-2870e6e85c76-c000.parquet
-rw-r--r--   3 hduser hadoop    4937730 2020-03-05 15:00 /data/petastorm_ingestion_tests/part-00003-daa9b6fc-7cf1-4404-aad5-2870e6e85c76-c000.parquet
-rw-r--r--   3 hduser hadoop    4911543 2020-03-05 15:00 /data/petastorm_ingestion_tests/part-00004-daa9b6fc-7cf1-4404-aad5-2870e6e85c76-c000.parquet
-rw-r--r--   3 hduser hadoop    3132042 2020-03-05 15:00 /data/petastorm_ingestion_tests/part-00005-daa9b6fc-7cf1-4404-aad5-2870e6e85c76-c000.parquet
-rw-r--r--   3 hduser hadoop    4816829 2020-03-05 15:00 /data/petastorm_ingestion_tests/part-00006-daa9b6fc-7cf1-4404-aad5-2870e6e85c76-c000.parquet
-rw-r--r--   3 hduser hadoop    4835624 2020-03-05 15:01 /data/petastorm_ingestion_tests/part-00007-daa9b6fc-7cf1-4404-aad5-2870e6e85c76-c000.parquet
-rw-r--r--   3 hduser hadoop    3187349 2020-03-05 15:01 /data/petastorm_ingestion_tests/part-00008-daa9b6fc-7cf1-4404-aad5-2870e6e85c76-c000.parquet
-rw-r--r--   3 hduser hadoop    6273385 2020-03-05 15:00 /data/petastorm_ingestion_tests/part-00009-daa9b6fc-7cf1-4404-aad5-2870e6e85c76-c000.parquet
```

**Environment**
petastorm==0.8.2
pyspark=2.4.4=py_0
libhdfs3=2.3.0=h2fca0e8_1

**Logs**
```
$ ./test_ingest.py
2020-03-05 13:29:29,067 WARN util.Utils: Your hostname, worker1 resolves to a loopback address: 127.0.1.1; using 10.x.x.x instead (on interface enp0s31f6)
2020-03-05 13:29:29,068 WARN util.Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
STAGE 1
STAGE 2
STAGE 3
2020-03-05 13:29:32,384 WARN hadoop.ParquetOutputFormat: Setting parquet.enable.summary-metadata is deprecated, please use parquet.summary.metadata.level
2020-03-05 13:29:44,585 WARN hadoop.ParquetOutputFormat: Setting parquet.enable.summary-metadata is deprecated, please use parquet.summary.metadata.level
DONE
Segmentation fault (core dumped)
```

**The code**
```python
#!/usr/bin/env python3

import os, sys
import numpy as np
from glob import glob
import cv2

from pyspark.sql import SparkSession, Row
from pyspark.sql.types import StructField, StructType, IntegerType, BinaryType, StringType, TimestampType

from petastorm.etl.dataset_metadata import materialize_dataset
from petastorm.codecs import ScalarCodec, CompressedImageCodec, NdarrayCodec
from petastorm.unischema import dict_to_spark_row, Unischema, UnischemaField

ROWGROUP_SIZE_MB = 128 # The same as the default HDFS block size

# The schema defines how the dataset schema looks like
ImageSchema = Unischema('ImageSchema', [
    UnischemaField('path', np.string_, (), ScalarCodec(StringType()), False),
    UnischemaField('image', np.uint8, (1080, 1280, 3), CompressedImageCodec('png'), False)
])


#output_url = "file:///tmp/petastorm_ingest_test"
output_url = "hdfs:///data/petastorm_ingestion_tests"
rows_count = 1

def row_generator(x):
    """Converts single row into a dictionary"""
    print("Creating row for: " + x.path)
    return {'path': x.path,
            'image': np.reshape(x.image, (1080,1280,3))}

def ingest_folder2(images_folder, spark):

    # List all images in the folder
    image_files = sorted(glob(os.path.join(images_folder, "*.png")))

    with materialize_dataset(spark, output_url, ImageSchema, ROWGROUP_SIZE_MB):

        print("STAGE 1")
        input_rdd = spark.sparkContext.parallelize(image_files) \
            .map(lambda image_path:
                    {ImageSchema.path.name: image_path,
                     ImageSchema.image.name: cv2.imread(image_path)})

        print("STAGE 2")
        rows_rdd = input_rdd.map(lambda r: dict_to_spark_row(ImageSchema, r))

        print("STAGE 3")
        spark.createDataFrame(rows_rdd, ImageSchema.as_spark_schema()) \
            .coalesce(10) \
            .write \
            .mode('overwrite') \
            .option('compression', 'none') \
            .parquet(output_url)

        print("DONE")

def main():

    # Start the Spark session
    spark = SparkSession.builder.config('spark.driver.memory', '2g').master('local[*]').getOrCreate()    
    sc = spark.sparkContext

    # Ingest images and annotations for a given folder
    ingest_folder2("../images", spark)
    print("Finished")

if __name__ == '__main__':
    main()
```

**Comment by selitvin** [comment]
1. Would the crash occur if you use libhdfs (and not libhdfs3) driver?
```python
resolver=FilesystemResolver(dataset_url, spark.sparkContext._jsc.hadoopConfiguration(), hdfs_driver='libhdfs')
with materialize_dataset(..., filesystem_factory=resolver.filesystem_factory()):
    ...
```
2. Do you know after which statement (probably somewhere in the `materialize_dataset`'s postamble code) the crash occurs?

**Comment by filipski** [comment]
Thanks again for quick reply!

1. This indeed helps. There's no more crash after switching from libhdfs3 to libhdfs, and the files seems to be generated properly, including `_common_metadata`

Logs:
```
$ ./test_ingest.py
2020-03-05 21:21:14,094 WARN util.Utils: Your hostname, worker1 resolves to a loopback address: 127.0.1.1; using 10.x.x.x instead (on interface enp0s31f6)
2020-03-05 21:21:14,094 WARN util.Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
STAGE 1
STAGE 2
STAGE 3
2020-03-05 21:21:18,386 WARN hadoop.ParquetOutputFormat: Setting parquet.enable.summary-metadata is deprecated, please use parquet.summary.metadata.level
2020-03-05 21:21:33,169 WARN hadoop.ParquetOutputFormat: Setting parquet.enable.summary-metadata is deprecated, please use parquet.summary.metadata.level
DONE
2020-03-05 21:21:33,381 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:33,427 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
/home/hduser/miniconda3/envs/petastorm/lib/python3.6/site-packages/petastorm/etl/dataset_metadata.py:192: FutureWarning: The 'add_metadata' method is deprecated, use 'with_metadata' instead
  utils.add_to_dataset_metadata(dataset, UNISCHEMA_KEY, serialized_schema)
2020-03-05 21:21:33,539 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
[Stage 1:>                                                        (0 + 10) / 10]2020-03-05 21:21:35,940 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:35,998 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,005 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,010 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,025 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,032 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,033 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,061 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,061 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,071 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,088 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,090 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,103 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,108 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,112 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,112 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,132 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,143 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,152 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,177 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
2020-03-05 21:21:36,307 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
/home/hduser/miniconda3/envs/petastorm/lib/python3.6/site-packages/petastorm/etl/dataset_metadata.py:228: FutureWarning: The 'add_metadata' method is deprecated, use 'with_metadata' instead
  utils.add_to_dataset_metadata(dataset, ROW_GROUPS_PER_FILE_KEY, num_row_groups_str)
2020-03-05 21:21:36,369 INFO sasl.SaslDataTransferClient: SASL encryption trust check: localHostTrusted = false, remoteHostTrusted = false
Finished
```

Folder content:
```
$ hdfs dfs -ls /data/petastorm_ingestion_tests
Found 12 items
-rw-r--r--   3 hduser hadoop          0 2020-03-05 21:26 /data/petastorm_ingestion_tests/_SUCCESS
-rw-r--r--   3 hduser hadoop       7206 2020-03-05 21:26 /data/petastorm_ingestion_tests/_common_metadata
-rw-r--r--   3 hduser hadoop    3204443 2020-03-05 21:26 /data/petastorm_ingestion_tests/part-00000-75d58b2f-ddf3-4e18-a682-9ce5e8a4a4fb-c000.parquet
-rw-r--r--   3 hduser hadoop    3267115 2020-03-05 21:26 /data/petastorm_ingestion_tests/part-00001-75d58b2f-ddf3-4e18-a682-9ce5e8a4a4fb-c000.parquet
-rw-r--r--   3 hduser hadoop    4943022 2020-03-05 21:26 /data/petastorm_ingestion_tests/part-00002-75d58b2f-ddf3-4e18-a682-9ce5e8a4a4fb-c000.parquet
-rw-r--r--   3 hduser hadoop    4937730 2020-03-05 21:26 /data/petastorm_ingestion_tests/part-00003-75d58b2f-ddf3-4e18-a682-9ce5e8a4a4fb-c000.parquet
-rw-r--r--   3 hduser hadoop    4911543 2020-03-05 21:26 /data/petastorm_ingestion_tests/part-00004-75d58b2f-ddf3-4e18-a682-9ce5e8a4a4fb-c000.parquet
-rw-r--r--   3 hduser hadoop    3132042 2020-03-05 21:26 /data/petastorm_ingestion_tests/part-00005-75d58b2f-ddf3-4e18-a682-9ce5e8a4a4fb-c000.parquet
-rw-r--r--   3 hduser hadoop    4816829 2020-03-05 21:26 /data/petastorm_ingestion_tests/part-00006-75d58b2f-ddf3-4e18-a682-9ce5e8a4a4fb-c000.parquet
-rw-r--r--   3 hduser hadoop    4835624 2020-03-05 21:26 /data/petastorm_ingestion_tests/part-00007-75d58b2f-ddf3-4e18-a682-9ce5e8a4a4fb-c000.parquet
-rw-r--r--   3 hduser hadoop    3187349 2020-03-05 21:26 /data/petastorm_ingestion_tests/part-00008-75d58b2f-ddf3-4e18-a682-9ce5e8a4a4fb-c000.parquet
-rw-r--r--   3 hduser hadoop    6273385 2020-03-05 21:26 /data/petastorm_ingestion_tests/part-00009-75d58b2f-ddf3-4e18-a682-9ce5e8a4a4fb-c000.parquet
```

2. I added a few print() statements to `materialize_dataset()` in `petastorm\etl\dataset_metadata.py` and it's clearly this piece of code which crashes:
```python
    dataset = pq.ParquetDataset(
        dataset_path,
        filesystem=filesystem,
        validate_schema=False)
```

So, what's your recommendation? Shall I stick to libhdfs? Shall we file a bug against libhdfs3 somewhere?

**Comment by selitvin** [comment]
Yep, not sure we can do much about this. I think arrow folks took over libhdfs3, but not sure how actively it is being maintained. If this is an option for you, it's probably best to just switch to libhdfs as it is well maintained and widely used.

I am closing the issue for now as it seems to be resolved.
